src/ProfileEvent.h

```diff
@@ -0,0 +1,358 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2018, The Souffle Developers. All rights reserved.
+ * Licensed under the Universal Permissive License v 1.0 as shown at:
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file ProfileEvent.h
+ *
+ * Declares classes for profile events
+ *
+ ***********************************************************************/
+
+#include <atomic>
+#include <chrono>
+#include <ctime>
+#include <functional>
+#include <iostream>
+#include <list>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+#include <assert.h>
+#include <sys/resource.h>
+#include <sys/time.h>
+
+#pragma once
+
+namespace souffle {
+
+typedef uint32_t ProfileKey;
+typedef uint32_t Interval;
+typedef std::chrono::system_clock::time_point ProfileTimePoint;
+
+/**
+ * ProfileKey Class
+ *
+ * Class converts text to keys for profile messages
+ * to reduce memory overheads.
+ */
+class ProfileKeySingleton {
+    std::unordered_map<std::string, ProfileKey> map;
+    std::vector<std::string> text;
+    std::mutex keyMutex;
+    ProfileKey last;
+
+private:
+    ProfileKeySingleton() : last(0){};
+
+public:
+    /** Get instance */
+    static ProfileKeySingleton& instance() {
+        static ProfileKeySingleton singleton;
+        return singleton;
+    }
+
+    /** Lookup text */
+    ProfileKey lookup(const std::string& txt) {
+        auto it = map.find(txt);
+        if (it != map.end()) {
+            return it->second;
+        } else {
+            std::lock_guard<std::mutex> guard(keyMutex);
+            map[txt] = last;
+            text.push_back(txt);
+            return last++;
+        }
+    }
+
+    /** Get text */
+    const std::string getText(ProfileKey key) {
+        std::lock_guard<std::mutex> guard(keyMutex);
+        assert(key < last && "corrupted key");
+        return text[key];
+    }
+};
+
+/** Abstract class for a profile event */
+class ProfileEvent {
+private:
+    /** Key of profile event */
+    ProfileKey key;
+
+    /** Start time of event */
+    ProfileTimePoint start;
+
+protected:
+    /** Convert ProfileTimePoint(time) to string */
+    const std::string timeToString(ProfileTimePoint& time) {
+        std::chrono::milliseconds ms =
+                std::chrono::duration_cast<std::chrono::milliseconds>(time.time_since_epoch());
+
+        std::chrono::seconds s = std::chrono::duration_cast<std::chrono::seconds>(ms);
+        std::time_t t = s.count();
+        std::size_t fractional_seconds = ms.count() % 1000;
+
+        char buffer[80];
```

Comment:
You could do this with C++ tools. Something like
```
std::stringstream ss;
ss << std::put_time(std::localtime(&time), "%F %T") << ":" << fractional_seconds;
return ss.str();
```
This can wait for a later PR if needed.